### PR TITLE
WIP: Automated testing

### LINKED ISSUE #26: Expanded automatic testing
Incorporate some automatic testing framework (pytest?) to run all the examples, and add addition edge case examples of things one might do using Java-Python bridge.

Since almost all of these tests would need to run with micro-manager already running, it would probably be difficult to have github do this automatically, but it should be feasible to have a single script that can be run locally to do it

### PR DESCRIPTION
**Pull request by ieivanov**
This PR sets up a GitHub Action that runs pytest on every pull request to the main branch. In pytest fixtures we download the latest MM nightly build and install it. So far, I have one test that starts headless mode and connects to the core. The test passes locally and in the GitHub Action virtual machine 🎉 

TODO
- [ ] Better cleanup - so far MM is installed, but the install directory is not removed at the end of the tests. I think MM is still running in headless mode at the time I try to delete the folder. As a workaround, I delete the install folder at the start of the tests if it exists. This allows us to run the test locally multiple times
- [x] Build PM java dependencies using maven and transfer jars to MM install directory as part of the GitHub Action build process (see #26)
- [ ] Transfer archived tests from `archive_test` to `pycromanager/test` with proper formatting

Fixed #26

### CODE REVIEW

**Comment by henrypinkard** [comment]
Awesome!!!

File: pycromanager/test/conftest.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+def launch_mm_headless(install_mm):
+    mm_install_dir = install_mm
+    config_file = os.path.join(mm_install_dir, 'MMConfig_demo.cfg')
+
+    print('Launching Micro-manager in headless mode.')
+    start_headless(mm_install_dir, config_file)
+
+    # yield None
+    #
+    # cleanup()
```
Comment (ieivanov) [inline pycromanager/test/conftest.py:118]: @henrypinkard can I get your help here? I need a way to shut down MM running in headless mode, but calling `acq_util.cleanup` is not doing it.

**Comment by ieivanov** [comment]
`test_zstack_seq_acq` is failing because of #461

File: pycromanager/test/conftest.py (hunk 1 above)
Comment (henrypinkard) [reply pycromanager/test/conftest.py:118]: Hmm weird. It seems to work for me (on mac). What OS are you on?

Comment (ieivanov) [reply pycromanager/test/conftest.py:138]: You're right, `cleanup()` does shut down MM, I just needed to give it a few seconds (`time.sleep(5)`) before `shutil.rmtree(mm_install_dir)` is called by the next function. However, these the `cleanup()` function (I think) does cause `pytest` to hang, bost locally and in Github Actions

Comment (henrypinkard) [reply pycromanager/test/conftest.py:139]: Hmm any ideas why? I'm happy to play around with it if you can point to a minimal example of this

Comment (ieivanov) [reply pycromanager/test/conftest.py]: Yeah, I can't figure it out...

I just modified the setup for the tests such that MM is installed only if it's not found in `~/Micro-Manager-nightly`. You should be able to run the test locally on your Mac if you place a working MM installation there. 

You'll find that if you uncomment these three lines pytest will hang up at the very end. The reason I want to properly shut down this headless instance is that I want to then test starting another one on a different port, or test starting two headless instances simultaneously.

Comment (henrypinkard) [reply pycromanager/test/conftest.py]: I just tried this on Mac and it didn't seem to cause problems. Seems like it must be a windows thing, which would make sense because there are differences in how Unix vs windows processes work. I can try also doing it on windows to see if I can get to the bottom of it

Comment (henrypinkard) [reply pycromanager/test/conftest.py]: Tried on windows too. All the tests pass and cleanup gets called as expected and it does indeed shutdown the Java process. I did manually install the latest nightly build. Maybe try that?

Comment (ieivanov) [reply pycromanager/test/conftest.py]: I still get the same behavior on Windows - the tests complete, but then pytests hangs up before fully exiting. This seems to be a problem in Github Actions too - the tests stall for 6 hours at which point they can cancelled: https://github.com/czbiohub/pycro-manager/actions/runs/3753840267.

For now, I've commented out the call to `cleanup()` again. Let's fix this in another PR

**Comment by ieivanov** [comment]
I think it's valuable to merge these tests soon, so we can avoid bugs. At present, pycromanager acquisitions don't work because `PycroManagerJava` doesn't compile - there seems to be a problem in the java code.

**Comment by henrypinkard** [comment]
Ok done. I'm just working through a big change related to and building upon https://github.com/micro-manager/NDTiffStorage/pull/69

Should be able to get everything tested and working again in the next day or two

**Comment by ieivanov** [comment]
Cleanup issues here were fixed by #548
